akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -444,108 +568,173 @@ private[akka] class Shard(
     case PassivateIdleTick                       => passivateIdleEntities()
     case msg: LeaseLost                          => receiveLeaseLost(msg)
     case msg: RememberEntityStoreCrashed         => rememberEntityStoreCrashed(msg)
-    case msg if extractEntityId.isDefinedAt(msg) => deliverMessage(msg, sender(), OptionVal.None)
+    case msg if extractEntityId.isDefinedAt(msg) => deliverMessage(msg, sender())
   }
 
-  def rememberRemove(entityId: String)(whenDone: () => Unit): Unit = {
+  def rememberUpdate(add: Set[EntityId] = Set.empty, remove: Set[EntityId] = Set.empty): Unit = {
     rememberEntitiesStore match {
       case None =>
-        whenDone()
+        onUpdateDone(add, remove)
       case Some(store) =>
-        flightRecorder.rememberEntityRemove(entityId)
-        sendToRememberStore(store, Set(entityId), RememberEntitiesShardStore.RemoveEntity(entityId))(whenDone)
-    }
-  }
-  def rememberAdd(entityIds: Set[EntityId])(whenDone: () => Unit): Unit = {
-    rememberEntitiesStore match {
-      case None =>
-        whenDone()
-      case Some(store) =>
-        entityIds.foreach { id =>
-          entities.remembering(id)
-          flightRecorder.rememberEntityAdd(id)
-        }
-        sendToRememberStore(store, entityIds, RememberEntitiesShardStore.AddEntities(entityIds))(whenDone)
+        sendToRememberStore(store, storingStarts = add, storingStops = remove)
     }
   }
 
   /**
    * The whenDone callback should not call become either directly or by calling this method again
    * as it uses become.
    */
-  def sendToRememberStore(store: ActorRef, entityIds: Set[EntityId], command: RememberEntitiesShardStore.Command)(
-      whenDone: () => Unit): Unit = {
-    if (VerboseDebug) log.debug("Update of [{}] [{}] triggered", entityIds.mkString(", "), command)
-    val startTime = System.nanoTime()
-    store ! command
+  def sendToRememberStore(store: ActorRef, storingStarts: Set[EntityId], storingStops: Set[EntityId]): Unit = {
+    if (VerboseDebug)
+      log.debug(
+        "Remember update [{}] and stops [{}] triggered",
+        storingStarts.mkString(", "),
+        storingStops.mkString(", "))
+
+    // FIXME, handle order of add remove when an id is in both somehow?
+    storingStarts.foreach { entityId =>
```

Comment:
perhaps this could be a precondition that an id is not in both?

Reply:
I just realised that is not how it can happen anymore, the precondition is already in place since it will require a state transition from RememberingStart => RememberingStop or RememberingStop => RememberingStart which currently will throw.

Reply:
let's remove the comment then